sycl/doc/extensions/experimental/sycl_ext_oneapi_enqueue_custom_operation.asciidoc

```diff
@@ -0,0 +1,184 @@
+is no need to do any further synchronization within the backend API in the user
+submitted lambda.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_ENQUEUE_CUSTOM_OPERATION` to one of the values defined
+in the table below.  Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+_And follow the text with a table like this *unless the extension is
+"experimental"*.  Note that your table may have more than one row if it
+has multiple versions._
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== Synchronization
+
+The goal of `sycl_ext_oneapi_enqueue_custom_operation` is to integrate interop
+work within the asynchronous SYCL DAG. As such, the user defined lambda must
+only enqueue asynchronous work within the user lambda. Asynchronous work must
+only be submitted to the native queue obtained from
+`interop_handle::get_native_queue`.
+
+A lambda submitted to `sycl_ext_oneapi_enqueue_custom_operation` may choose to
+not wait on dependent events to complete, in order for asynchronous interop
+work to be composed with asynchronous SYCL commands. This gives different
+sychronization guarantees from normal SYCL `host_task`s.
+
+In this example:
+
+```
+q.submit([&](sycl::handler &cgh) {
+    cgh.depends_on(dep_event);
+    cgh.sycl_ext_oneapi_enqueue_custom_operation([=](sycl::interop_handle &h) {
+      printf("This will print before dep_event has completed.\n");
+      hipStream_t stream = h.get_native_queue<sycl::backend::hip>();
+      hipMemcpyAsync(target_ptr, native_mem, test_size * sizeof(int),
+                      hipMemcpyDeviceToHost, stream);
+    });
+  });
+q.wait()
+```
+
+The print statement may print before `dep_event` has completed. However, the
+asynchronous memcpy submitted to the native queue obtained by
+`interop_handle::get_native_queue` is guaranteed to have the correct
+dependencies, and therefore will only start once its dependent events have
+completed.
+
+By contrast, when using a `host_task`, it is guaranteed that the print statment
+will only happen once the host task's dependent events are observed to be
+complete on the host.
+
+
+=== Additions to handler class
+
+This extension adds the following new member function to the SYCL `handler`
+class:
+
+```
+class handler {
+  template <typename Func>
+  void sycl_ext_oneapi_enqueue_custom_operation(Func&& interopCallable);
+};
+```
+
+`interopCallable` must be a callable which takes a `interop_handle` as its
+first arg.
```

Comment:
We are trying to be more precise about requirements like this.  What exactly does "must be a callable..." mean?  In this case, I think this is a constraint on the template parameter `Func`.  If so, we should document it like:

> _Constraints:_ The `Func` must a C++ callable object which takes a single parameter of type `interop_handle`.

Reply:
Thanks have changed.